subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchableHierarchies.java

```diff
@@ -67,12 +69,27 @@ public void registerWatchableHierarchy(File watchableHierarchy, SnapshotHierarch
     @CheckReturnValue
     public SnapshotHierarchy removeUnwatchedSnapshots(Predicate<File> isWatchedHierarchy, SnapshotHierarchy root, Invalidator invalidator) {
         recentlyUsedHierarchies.removeIf(hierarchy -> !isWatchedHierarchy.test(hierarchy));
+        SnapshotHierarchy currentRoot = stopWatchingHierarchiesOverLimit(invalidator, root);
+
         this.watchableHierarchies = DefaultFileHierarchySet.of(recentlyUsedHierarchies);
-        RemoveUnwatchedFiles removeUnwatchedFilesVisitor = new RemoveUnwatchedFiles(root, invalidator);
-        root.visitSnapshotRoots(snapshotRoot -> snapshotRoot.accept(removeUnwatchedFilesVisitor));
+        RemoveUnwatchedFiles removeUnwatchedFilesVisitor = new RemoveUnwatchedFiles(currentRoot, invalidator);
+        currentRoot.visitSnapshotRoots(snapshotRoot -> snapshotRoot.accept(removeUnwatchedFilesVisitor));
         return removeUnwatchedFilesVisitor.getRootWithUnwatchedFilesRemoved();
     }
 
+    private SnapshotHierarchy stopWatchingHierarchiesOverLimit(Invalidator invalidator, SnapshotHierarchy root) {
+        SnapshotHierarchy result = root;
+        int toRemove = recentlyUsedHierarchies.size() - maxHierarchiesToWatch;
+        if (toRemove > 0) {
+            LOGGER.warn("Watching {} hierarchies, removing {} to limit the number of watched directories", recentlyUsedHierarchies.size(), toRemove);
```

Comment:
Should this appear on the console?

If so, I think we need to give it more context so that it can be easily understood on its own. Something like:

```suggestion
            LOGGER.warn("Watching too many directories in the file system (watching {}, limit {}), dropping some state from the virtual file system", ...);
```

Reply:
I think we should warn people when they hit the limit, so that they can take action and inform us about it.

Reply:
I adjusted the message so it gives more context.